### ISSUE TITLE
Truncated menu and how to fix it

### ISSUE THREAD

**Issue by SteveTheTechie** [ehynds/jquery-ui-multiselect-widget#739]
In some cases, the menu can appear truncated.  In particular, you will see this if the multiselect is located in a jQuery UI dialog.   It has to do with the overflow settings of the parent container, what the appendTo is set to, and how it is positioned.

This is an issue that has been addressed a number of years ago by the developers of Chosen (w/ fixes):
https://github.com/harvesthq/chosen/issues/86

**Comment by SteveTheTechie** [comment]
Also relevant:
[https://github.com/harvesthq/chosen/pull/290/commits/a53b58f15a3734ca3affe7c1fe355335aa7d09e8](https://github.com/harvesthq/chosen/pull/290/commits/a53b58f15a3734ca3affe7c1fe355335aa7d09e8)

**Comment by SteveTheTechie** [comment]
The following shows this issue.  It is a small jQuery UI Modal Dialog, and the menu is getting clipped instead of overlaying the dialog like a native select does.
<img width="236" alt="capture-ms-clip" src="https://user-images.githubusercontent.com/5685883/36015325-ea0ea002-0d22-11e8-8c23-6d02090d2b4e.PNG">

**Comment by SteveTheTechie** [comment]
@mlh758 The following shows this issue.  It is a small jQuery UI Modal Dialog, and the menu is getting clipped instead of overlaying the dialog like a native select does.

<img width="236" alt="capture-ms-clip" src="https://user-images.githubusercontent.com/5685883/36015325-ea0ea002-0d22-11e8-8c23-6d02090d2b4e.PNG">

**Comment by SteveTheTechie** [comment]
Closing this issue.

**Comment by mlh758** [comment]
Fascinating, I switched the append behavior to what it is now because it used to pretty much always go to the body. When the widget was set in a dialog it couldn't be interacted with at all, even if it appeared over it.

Could you add this example to the wiki somewhere?

**Comment by SteveTheTechie** [comment]
Sure.  The problem is that we need position:absolute for the menu container so that it is not constrained to the dialog--that is in the CSS already.  The position() _function_ does the magic of positioning the menu to the button, which does reside in the container.   Also, since position:absolute removes the element from the "flow", it does not take up any space in the layout... it is sort of a "free agent", _but only within the constraints of what you are appending to_.

**Comment by SteveTheTechie** [comment]
@mlh758  I am not sure how much value the following code is in `_getAppendEl()`, since it likely causes the menu to get truncated in some situations (e.g. dialogs w/ short heights):
```
if (!elem || !elem[0]) {
   elem = this.element.closest(".ui-front, dialog");
}
```

**Comment by mlh758** [comment]
I used it because it mirrors the default stacking order in jQuery UI. I'll
have to find you the link, it's specified in their docs somewhere and was
implemented because it resolved issues with the menu not being clickable
inside of dialogs. It also allowed me to remove some heavy z index styles
from the CSS.

**Comment by mlh758** [comment]
https://api.jqueryui.com/theming/stacking-elements/

**Comment by SteveTheTechie** [comment]
@mlh758 Ok, I am reopening this as it appears there is more to discuss on this...  

With all due respect, don't put so much stock in the jQuery UI documentation.  I have personally found that much of their code does not quite work as advertised or their documentation is a bit confusing at times.  It can be very frustrating at times.

There are several things going on here and the referenced doc does not do an adequate job of explaining it in my view.  

1.  I could see how you might get the impression that appending to the element w/ .ui-front would handle the z-index for you.  However, that is not really the whole story and the doc does not cover some of the related issues.

2. Because the menu container must be position: absolute to work properly as a overlay, its display is limited to the bounds of the parent element it is appended to if the parent element has overflow:hidden set.   https://stackoverflow.com/questions/14837553/overflowhidden-does-not-clip-absolute-positioned-content _(kind of the converse argument, but still helpful)_

3. Because .ui-front frequently corresponds to the outer div container of a jQuery UI dialog that always has `overflow: hidden` set, this means that the menu will get clipped if it extends outside of the parent div, _regardless of z-index_.   Of course, they do not really mention this in the referenced doc. 👎  This is **always** going to be a problem for this widget when you append to an element that has `overflow: hidden` set.

4. Additionally, appending to .ui-front does not really set the z-index... it just keeps it the same as what ever .ui-front is... sometimes that is insufficient if there are other overlay elements in the dialog.

Here is what to do:
---
1.  Disqualify elements in getAppendEl() that have `overflow: hidden` set to prevent the clipping issue.
2.  Add an option for zIndex to be explicitly set if need be.  If it is null, then determine a best z-Index as the (maximum the z-indexes of closest .ui-front element, the getAppendEl) + 1 .   This will keep the z-index manageable while setting it explicitly.  
- Alternatively, we could create a zIndex option with a default value of 120 (probably sufficient)--that will overlay all jQuery UI elements except for tool tips (z-index=9999).
3.   Since it looks like appendTo could be set to the body more frequently, I will need to ditch the cached state of position() to force the menu to get repositioned on every open in case the button is moved from the last open (e.g dialog is moved).  This will be necessary if the parent container of the menu is the body not the dialog.

**Comment by SteveTheTechie** [comment]
> I used it because it mirrors the default stacking order in jQuery UI. I'll
> have to find you the link, it's specified in their docs somewhere and was
> implemented because it resolved issues with the menu not being clickable
> inside of dialogs. It also allowed me to remove some heavy z index styles
> from the CSS.

Ok, but the proper fix is a jQuery UI dialog extension point called _allowInteraction()...  It is on the jQuery UI dialog page.  https://api.jqueryui.com/dialog/  The reason I know this is because I have had to use _allowInteraction)_ w/ CKEditor() instances I put inside modal dialogs.

**Comment by mlh758** [comment]
> Additionally, appending to .ui-front does not really set the z-index... it just keeps it the same as what ever .ui-front is... sometimes that is insufficient if there are other overlay elements in the dialog.

The point was to not have to set any z-index, the menu is properly inside of the overlay element and can be interacted with, without any need for z-index. There used to be a hard coded z-index of 10,000 in the css that I got rid of with this change because just setting a z-index value puts you in an inevitable z-index war with other ui tools.

I get that your suggestion is to make that z-index dynamic in JavaScript instead but I'm not immediately convinced that going back to body as default with a z-index, against the documentation's recommendation, is the best route for a fix given the number of issues that were logged to this widget for a similar setup before.

> Since it looks like appendTo could be set to the body more frequently, I will need to ditch the cached state of position() to force the menu to get repositioned on every open in case the button is moved from the last open (e.g dialog is moved). This will be necessary if the parent container of the menu is the body not the dialog.

This is probably a good idea no matter what we end up doing for a fix.

jQuery Autocomplete seems to use the same method I am using. Select2 defaults to the body. I'll look around a bit.

**Comment by mlh758** [comment]
jQuery Autocomplete suffers from the truncation issue.

**Comment by mlh758** [comment]
Just to clarify, I'm not saying no to your proposal, I just want to look around a bit myself before moving away from the jQuery UI recommendation on stacking order.

**Comment by SteveTheTechie** [comment]
> jQuery Autocomplete suffers from the truncation issue.

I bet if you set its appendTo option to document.body, it does not get truncated... however, its menu may show up behind a dialog for that case, since the body would be expected to have a lower z-index.

This illustrates why I think this widget will need some option to explicitly set the z-index for the cases where the appendTo is the document.body.  Alternatively, if the button's parent has class ui-front, we could force an appropriate z-index for that case.

**Comment by SteveTheTechie** [comment]
> Just to clarify, I'm not saying no to your proposal, I just want to look around a bit myself before moving away from the jQuery UI recommendation on stacking order.

Understood.  I thought there was more to consider... that is why I reopened this issue.

**Comment by SteveTheTechie** [comment]
@mlh758 
#### I will go ahead and go back to setting position on every open.

To summarize:
1.  The z-index will only get handled for the case of appending to a container w/ class ui-front.  However, even for this case, there are situations where you may need to override the z-index used for the menu.  (e. g.,  if you have multiple elements that have overlay elements)
2.  The ui-front approach allows for the menu to be truncated if the parent container is short and has overflow: hidden set as any jQuery UI dialog would have.
3.  If you instead append to document.body, you fix the truncation issue, but you may have to explicitly set the z-index and you may have interaction issues w/ jQuery modal dialogs, that requires the use of the _allowInteraction extension point to white list the menu container.  The reason for this is that if you append to document.body, then the menu is no longer a child element of the jQuery modal dialog, so it the modal dialog code will block interaction w/ it unless you specifically tell the code not to.
4.  If another element is specified as the appendTo option, there may still be a need to explicitly set the z-index for the menu on a case by case basis.  We could add a fall-back z-index in the CSS, but that will be global, not case by case like an option.  Also, interaction in modal dialogs could still be an issue.

**Comment by SteveTheTechie** [comment]
Example of _allowInteraction usage:
```
jQuery.widget( "ui.dialog", jQuery.ui.dialog, {
	 /*! jQuery UI - v1.10.2 - 2013-12-12
	  *  http://bugs.jqueryui.com/ticket/9087#comment:27 - bugfix
	  *  http://bugs.jqueryui.com/ticket/4727#comment:23 - bugfix
	  *  allowInteraction fix to accommodate windowed editors
	  */
	_allowInteraction: function( event ) {
		 if ( this._super( event ) ) {
			return true;
		 }

		 // address interaction issues with general iframes with the dialog
		 if ( event.target.ownerDocument != this.document[ 0 ] ) {
			return true;
		 }

		 // address interaction issues with dialog window
		 if ( jQuery( event.target ).closest( ".cke_dialog" ).length ) {
			return true;
		 }

		 // address interaction issues with iframe based drop downs in IE
		 if ( jQuery( event.target ).closest( ".cke" ).length ) {
			return true;
		 }
		 
		 if ( jQuery( event.target ).closest( ".sp-input" ).length ) {
			return true;
		 }
	},
	/*! jQuery UI - v1.10.2 - 2013-10-28
	*  http://dev.ckeditor.com/ticket/10269 - bugfix
	*  moveToTop fix to accommodate windowed editors
	*/
	_moveToTop: function ( event, silent ) {
		 if ( !event || !this.options.modal ) {
			this._super( event, silent );
		 }
	}
});
```
Extending widgets reference (related to above): https://learn.jquery.com/jquery-ui/widget-factory/extending-widgets/

**Comment by SteveTheTechie** [comment]
Addressed by #749

**Comment by SteveTheTechie** [comment]
@mlh758 Some additional thoughts recorded for "posterity"...

1. It turns out that there is no perfect approach to deal with this problem.
2. If you append to an element w/ `.ui-front set`, you may not need to set the checkbox container z-index or interaction issues, but you may get a truncated menu because of `overflow: hidden` being set on the container you are appending to.
3. If you instead append to the body, you get rid of the truncated menu problem, but you may have z-index or interaction issues, since the jQuery dialog code would then not recognize the menu as a descendant and would try to put it behind the dialog and prevent interaction in the case of modal dialogs.
4. We could test whether the element we are appending to has `overflow:hidden` set and switch to appending to the body if it does, but as I tend to believe that most jQuery dialog containers probably have that set, this would have the effect of essentially causing them never to get appended to.
5. This is why the best that we can probably do is make the widget work well for whatever element it is appended to, including the scenario of it being appended to the body and potentially having interaction issues with modals.
